src/IB/CIBStochasticMethod.cpp

```diff
@@ -0,0 +1,463 @@
+    }
+}
+
+void
+CIBStochasticMethod::checkLagUpdate()
+{
+    int reject = 0;
+
+    Eigen::Matrix3d rotation_mat_new = Eigen::Matrix3d::Zero();
+    Eigen::Vector3d dr = Eigen::Vector3d::Zero();
+    Eigen::Vector3d R_dr_new = Eigen::Vector3d::Zero();
+
+    Eigen::Vector3d X_new = Eigen::Vector3d::Zero();
+
+    // Get the grid extents.
+    Pointer<CartesianGridGeometry<NDIM> > grid_geom = d_hierarchy->getGridGeometry();
+
+    const double* const domain_x_lower = grid_geom->getXLower();
+    const double* const domain_x_upper = grid_geom->getXUpper();
+    double domain_length[NDIM];
+    for (int d = 0; d < NDIM; ++d)
+    {
+        domain_length[d] = domain_x_upper[d] - domain_x_lower[d];
+    }
+    const IntVector<NDIM>& periodic_shift = grid_geom->getPeriodicShift();
+
+    // Rotate the body with new rotational velocity about origin
+    // and translate the body to newer position.
+    const int coarsest_ln = 0;
+    const int finest_ln = d_hierarchy->getFinestLevelNumber();
+    for (int ln = coarsest_ln; ln <= finest_ln; ++ln)
+    {
+        if (!d_l_data_manager->levelContainsLagrangianData(ln)) continue;
+
+        const boost::multi_array_ref<double, 2>& X0_array =
+            *(d_l_data_manager->getLData("X0_unshifted", ln)->getLocalFormVecArray());
+        const Pointer<LMesh> mesh = d_l_data_manager->getLMesh(ln);
+        const std::vector<LNode*>& local_nodes = mesh->getLocalNodes();
+
+        // Get structures on this level.
+        const std::vector<int> structIDs = d_l_data_manager->getLagrangianStructureIDs(ln);
+        const unsigned structs_on_this_ln = (unsigned)structIDs.size();
+#if !defined(NDEBUG)
+        TBOX_ASSERT(structs_on_this_ln == d_num_rigid_parts);
+#endif
+
+        for (std::vector<LNode*>::const_iterator cit = local_nodes.begin(); cit != local_nodes.end(); ++cit)
+        {
+            const LNode* const node_idx = *cit;
+            const int lag_idx = node_idx->getLagrangianIndex();
+            const int local_idx = node_idx->getLocalPETScIndex();
+            const double* const X0 = &X0_array[local_idx][0];
+
+            int struct_handle = 0;
+            if (structs_on_this_ln > 1) struct_handle = getStructureHandle(lag_idx);
+
+            for (unsigned int d = 0; d < NDIM; ++d)
+            {
+                dr[d] = X0[d] - d_center_of_mass_initial[struct_handle][d];
+            }
+
+            // Rotate dr vector using the rotation matrix.
+            rotation_mat_new = d_quaternion_new[struct_handle].toRotationMatrix();
+            R_dr_new = rotation_mat_new * dr;
+
+            for (unsigned int d = 0; d < NDIM; ++d)
+            {
+                X_new(d) = d_center_of_mass_new[struct_handle][d] + R_dr_new(d);
+                if (!periodic_shift[d])
+                {
+                    if (X_new(d) < domain_x_lower[d] || domain_x_upper[d] < X_new(d))
+                    {
+                        reject++;
+                        pout << "REJECTED BY WALL\n";
```

Comment:
We should either guard this with `d_enable_logging` (so that it can be conditionally turned off) and send it to `plog` or remove it.

Reply:
```suggestion
                       if (d_enable_logging) pout << "provisional time step rejected because of particle collision with a wall\n";
```